device_train.py

```diff
@@ -0,0 +1,288 @@
+    args = parse_args()
+    params = json.load(open(args.config))
+
+    gradient_accumulation_steps = params.get("gradient_accumulation_steps", 1)
+    per_replica_batch = params["per_replica_batch"]
+    cores_per_replica = params["cores_per_replica"]
+
+    assert cores_per_replica <= 8
+
+    bucket = params["bucket"]
+    model_dir = params["model_dir"]
+    layers = params["layers"]
+    d_model = params["d_model"]
+    n_heads = params["n_heads"]
+    n_vocab = params["n_vocab"]
+    seq = params["seq"]
+    norm = params["norm"]
+
+    val_batches = params["val_batches"]
+    val_every = params["val_every"]
+    ckpt_every = params["ckpt_every"]
+    keep_every = params["keep_every"]
+    eval_tasks = params["eval_harness_tasks"]
+    total_steps = params["total_steps"]
+
+    pe = params["pe"]
+    assert pe in ["fixed", "rotary", "t5"]
+
+    warmup_steps = params["warmup_steps"]
+    anneal_steps = params["anneal_steps"]
+    lr = params["lr"]
+    end_lr = params["end_lr"]
+    weight_decay = params["weight_decay"]
+
+    opt = optax.chain(
+        optax.scale(1 / gradient_accumulation_steps),
+        clip_by_global_norm(1),
+        optax.scale_by_adam(),
+        additive_weight_decay(weight_decay),
+        optax.scale(-1),
+        optax.scale_by_schedule(util.gpt3_schedule(warmup_steps, anneal_steps, lr, end_lr))
+    )
+
+    params["optimizer"] = opt
+
+    start = time.time()
+    print(f"jax devices: {jax.device_count()}")
+    print(f"jax runtime initialized in {time.time() - start:.06}s")
+
+    mesh_shape = (jax.device_count() // cores_per_replica, cores_per_replica)
+    devices = np.array(jax.devices()).reshape(mesh_shape)
+
+    # pick initial ckpt - based on tuning vs train from scratch
+
+    step = 0
+    initial_ckpt_state_path = None
+    train_loader = None
+
+    if args.tune_model_path:
+        initial_ckpt_state_path = args.tune_model_path
+        print('we are fine-tuning')
+    else:
+        print('we are not fine-tuning')
+        initial_ckpt_model_dir = model_dir
+        initial_ckpt_path = f"gs://{bucket}/{initial_ckpt_model_dir}"
+        meta_path = f"{initial_ckpt_path}/meta.json"
+
+        try:
+            with open(meta_path, "r") as f:
+                meta = json.load(f)
+            ckpt_step = meta["checkpoints"][-1]
+            initial_ckpt_state_path = f"{initial_ckpt_path}/step_{ckpt_step}/"
+            print(f"state will be restored from checkpoint {ckpt_step}")
+
+            step = ckpt_step
+            train_loader = meta['aux'].get("train_loader", None)
+        except NotFound:
+            # no checkpoint, start at zero
+            print(f"No checkpoint to load at {initial_ckpt_path}. Training from scratch.")
+
+    if initial_ckpt_state_path:
+        print(f"path to load checkpoint from: {initial_ckpt_state_path}")
+    else:
+        print("not loading from a checkpoint")
+
+    # set up datasets
+    print("setting up datasets")
+    tpu_size = jax.device_count()
```

Comment:
you should probs check that tpu_size <= cores per replica here

Reply:
What situation would cause this check to fail?

(Trying to understand what I'm preventing so I can, e.g., write an error message about it)

Reply:
sometimes JAX doesn't detect the TPUs properly (and detects a single CPU device), or if there is a future model with more than 8 shards

Reply:
Doesn't that mean we check that tpu_size >= cores per replica, not <=?

Reply:
Oh yes, my mistake